README


This archive contains the distribution Facebook-Graph-Cmdline,
version 0.123480:

  Extends Facebook::Graph with embedded HTTP for Oauth and token saving

This software is copyright (c) 2012 by Andrew Grangaard.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Facebook::Graph::Cmdline

Changes:
0.123482  2012-12-13 11:41:19 America/Los_Angeles
  * renames roles to better match naming scheme in Facebook::Graph
    role/HTTPtoken  => Cmdline/Role/HTTPAccessToken
    role/save_token => Cmdline/Role/SaveAccessToken
  * uses access_token() from parent Facebook::Graph rather than a parallel
    token().  Augments access_token to add lazybuild.
  * improves examples
  * initial tests under Moose and Mouse
  * adds an app id and secret to the example/facebook.yml
  * source build generates a README.mkdn for github/etc with coderwall!

0.123460  2012-12-10
  * release for la.pm.org dec 11 meeting.
  * initial functionality


lib/Facebook/Graph/Cmdline.pm
package Facebook::Graph::Cmdline;
{
  $Facebook::Graph::Cmdline::VERSION = '0.123482';
}

#ABSTRACT: Extends Facebook::Graph with embedded HTTP for Oauth and token saving

use Any::Moose;
use v5.10;

extends 'Facebook::Graph';
with 'Facebook::Graph::Cmdline::Role::HTTPAccessToken';

#Is there a better way to do MooseX vs MouseX 'with' loading?
#can import with "use Any::Moose 'X::SimpleConfig'" but that doesn't
#provide the action of "with," Mo*se::Util::apply_all_roles()
if (Any::Moose::moose_is_preferred)
{
    with 'MooseX::SimpleConfig';
    with 'MooseX::Getopt';
}
else
{
    with 'MouseX::SimpleConfig';
    with 'MouseX::Getopt';
}
# requires provided by ::SimpleConfig
with 'Facebook::Graph::Cmdline::Role::SaveAccessToken';

no Any::Moose;
__PACKAGE__->meta->make_immutable;

__END__
=pod

=head1 NAME

Facebook::Graph::Cmdline - Extends Facebook::Graph with embedded HTTP for Oauth and token saving

=head1 VERSION

version 0.123482

=head1 AUTHOR

Andrew Grangaard <spazm@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Andrew Grangaard.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Facebook/Graph/role/HTTPtoken.pm
package Facebook::Graph::role::HTTPtoken;
{
  $Facebook::Graph::role::HTTPtoken::VERSION = '0.123480';
}

#ABSTRACT: Embeds an HTTP::Daemon to implement OAuth callback for Facebook Authorization of Commandline Facebook apps.

use v5.10;
use Any::Moose 'Role';

#all provided by Facebook::Graph
requires qw(
    access_token
    authorize
    fetch
    postback
    request_access_token
);

has +postback => ( is => 'ro', required => 1 );

use HTTP::Daemon 6.00;
use URI;

###
# provides code, token
# requires permissions
# can override prompt_message, success_message

has code => (
    is         => 'rw',
    lazy_build => 1,
);
has token => (
    is         => 'rw',
    lazy_build => 1,
);
has permissions => (
    is      => 'ro',
    default => sub { [] }
);

# fmt will be called with url as arg
has prompt_message_fmt => (
    is      => 'rw',
    default => "Please visit this url to authorize application:\n%s\n"
);
has success_message => (
    is      => 'rw',
    default => 'Success!'
);

sub _build_code
{
    my $self = shift;
    my $uri  = $self
        ->authorize
        ->extend_permissions( @{ $self->permissions } )
        ->uri_as_string;
    printf $self->prompt_message_fmt, $uri;

    use HTTP::Daemon;
    my $postback = URI->new( $self->postback );
    my $d        = HTTP::Daemon->new(
        LocalAddr => $postback->host,
        LocalPort => $postback->port,
    ) || die;

    my $code = '';
    until ($code)
    {
        my $c = $d->accept;
        my $r = $c->get_request;
        next unless $r;

        if (    $r->url->path eq $postback->path
            and $r->url->query_param('code') )
        {
            $code = $r->url->query_param('code');
            $c->send_response(
                HTTP::Response->new(
                    200, undef, undef, $self->success_message
                )
            );
        }
        else
        {
            $c->send_response('204');
        }
    }
    $code;
}

sub _build_token
{
    my $self = shift;
    return $self->request_access_token( $self->code )->token;
}

sub verify_token
{
    my $self = shift;
    return 0 unless $self->has_token();

    say "verifying token";    ## DEBUG
    $self->access_token( $self->token );
    my $resp;
    eval { $resp = $self->fetch('me') };
    if ($@)
    {
        say "Bad access token, deleting";    ## INFO
        $self->clear_token;
        return 0;
    }
    return 1;
}

1;

__END__
=pod

=head1 NAME

Facebook::Graph::role::HTTPtoken - Embeds an HTTP::Daemon to implement OAuth callback for Facebook Authorization of Commandline Facebook apps.

=head1 VERSION

version 0.123480

=head1 AUTHOR

Andrew Grangaard <spazm@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Andrew Grangaard.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Facebook/Graph/role/save_token.pm
package Facebook::Graph::role::save_token;
{
  $Facebook::Graph::role::save_token::VERSION = '0.123480';
}

#ABSTRACT: Provides a save_token method to save token value to YAML config file.

use v5.10;
use Any::Moose 'Role';
use YAML::Any;

requires qw( get_config_from_file configfile token );
# MooseX::SimpleConfig or MouseX::SimpleConfig will provide
# both get_config_from_file and configfile


sub save_token
{
    my $self = shift;
    if ( !$self->configfile )
    {
        say "please save token: " . $self->token;
        return 1;
    }

    my $config = $self->get_config_from_file( $self->configfile );
    if ( !exists $config->{token} or $self->token ne $config->{token} )
    {
        $config->{token} = $self->token;
        YAML::Any::DumpFile( $self->configfile, $config );
    }
}

1;

__END__
=pod

=head1 NAME

Facebook::Graph::role::save_token - Provides a save_token method to save token value to YAML config file.

=head1 VERSION

version 0.123480

=head1 METHODS

=head2 save_token

Updates token value in configfile and saves as YAML if modified.

If configfile is not defined, the token is printed to STDOUT for manual saving.

=head1 AUTHOR

Andrew Grangaard <spazm@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Andrew Grangaard.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Facebook/Graph/Cmdline/Role/HTTPAccessToken.pm
package Facebook::Graph::Cmdline::Role::HTTPAccessToken;
{
  $Facebook::Graph::Cmdline::Role::HTTPAccessToken::VERSION = '0.123482';
}

#ABSTRACT: Embeds an HTTP::Daemon to implement OAuth callback for Facebook Authorization of Commandline Facebook apps.

use v5.10;
use Any::Moose 'Role';

#all provided by Facebook::Graph
requires qw(
    access_token
    authorize
    fetch
    postback
    request_access_token
);

has +postback     => ( is => 'ro', required   => 1 );
has +access_token => ( is => 'rw', lazy_build => 1 );

use HTTP::Daemon 6.00;
use URI;

###
# provides code, token
# requires permissions
# can override prompt_message, success_message

has code => (
    is         => 'rw',
    lazy_build => 1,
);
#has token => (
#    is         => 'rw',
#    lazy_build => 1,
#);
has permissions => (
    is      => 'ro',
    default => sub { [] }
);

# fmt will be called with url as arg
has prompt_message_fmt => (
    is      => 'rw',
    default => "Please visit this url to authorize application:\n%s\n"
);
has success_message => (
    is      => 'rw',
    default => 'Success!'
);

sub _build_code
{
    my $self = shift;
    my $uri  = $self
        ->authorize
        ->extend_permissions( @{ $self->permissions } )
        ->uri_as_string;
    printf $self->prompt_message_fmt, $uri;

    use HTTP::Daemon;
    my $postback = URI->new( $self->postback );
    my $d        = HTTP::Daemon->new(
        LocalAddr => $postback->host,
        LocalPort => $postback->port,
    ) || die;

    my $code = '';
    until ($code)
    {
        my $c = $d->accept;
        my $r = $c->get_request;
        next unless $r;

        if (    $r->url->path eq $postback->path
            and $r->url->query_param('code') )
        {
            $code = $r->url->query_param('code');
            $c->send_response(
                HTTP::Response->new(
                    200, undef, undef, $self->success_message
                )
            );
        }
        else
        {
            $c->send_response('204');
        }
    }
    $code;
}

sub _build_access_token
{
    my $self = shift;
    return $self->request_access_token( $self->code )->token;
}

sub verify_access_token
{
    my $self = shift;
    return 0 unless $self->has_access_token();

    say "verifying token";    ## DEBUG
    #$self->access_token( $self->token );
    my $resp;
    eval { $resp = $self->fetch('me') };
    if ($@)
    {
        say "Bad access_token, deleting";    ## INFO
        $self->clear_access_token;
        return 0;
    }
    return 1;
}

1;

__END__
=pod

=head1 NAME

Facebook::Graph::Cmdline::Role::HTTPAccessToken - Embeds an HTTP::Daemon to implement OAuth callback for Facebook Authorization of Commandline Facebook apps.

=head1 VERSION

version 0.123482

=head1 AUTHOR

Andrew Grangaard <spazm@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Andrew Grangaard.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Facebook/Graph/Cmdline/Role/SaveAccessToken.pm
package Facebook::Graph::Cmdline::Role::SaveAccessToken;
{
  $Facebook::Graph::Cmdline::Role::SaveAccessToken::VERSION = '0.123482';
}

#ABSTRACT: Provides a save_access_token method to save token value to YAML config file.

use v5.10;
use Any::Moose 'Role';
use YAML::Any;

requires qw( get_config_from_file configfile access_token );

# MooseX::SimpleConfig or MouseX::SimpleConfig will provide
# both get_config_from_file and configfile


sub save_access_token
{
    my $self = shift;
    if ( !$self->configfile )
    {
        say "please save token: " . $self->access_token;
        return 1;
    }

    my $config = $self->get_config_from_file( $self->configfile );
    if ( !exists $config->{access_token}
        or $self->access_token ne $config->{access_token} )
    {
        $config->{access_token} = $self->access_token;
        YAML::Any::DumpFile( $self->configfile, $config );
    }
}

1;

__END__
=pod

=head1 NAME

Facebook::Graph::Cmdline::Role::SaveAccessToken - Provides a save_access_token method to save token value to YAML config file.

=head1 VERSION

version 0.123482

=head1 METHODS

=head2 save_access_token

Updates token value in configfile and saves as YAML if modified.

If configfile is not defined, the token is printed to STDOUT for manual saving.

=head1 AUTHOR

Andrew Grangaard <spazm@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Andrew Grangaard.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Extends Facebook::Graph with embedded HTTP for Oauth and token saving",
  "AUTHOR" => "Andrew Grangaard <spazm\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Moose" => 0,
    "Mouse" => 0,
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Facebook-Graph-Cmdline",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Facebook::Graph::Cmdline",
  "PREREQ_PM" => {
    "Any::Moose" => 0,
    "Facebook::Graph" => 0,
    "HTTP::Daemon" => "6.00",
    "MooseX::Getopt" => 0,
    "MooseX::SimpleConfig" => 0,
    "MouseX::Getopt" => 0,
    "MouseX::SimpleConfig" => 0,
    "URI" => 0,
    "YAML::Any" => 0
  },
  "VERSION" => "0.123482",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



